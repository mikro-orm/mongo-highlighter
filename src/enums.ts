export enum TokenType {
  WHITESPACE,
  WORD,
  QUOTE,
  BACKTICK_QUOTE,
  BOUNDARY,
  NUMBER,
  ERROR,
  VARIABLE,
  LITERAL,
}

export enum HighlightSubject {
  BOUNDARY = 'boundary',
  WORD = 'word',
  BACKTICK_QUOTE = 'backtickQuote',
  QUOTE = 'quote',
  NUMBER = 'number',
  VARIABLE = 'variable',
  LITERAL = 'literal',
}

export const TOKEN_TYPE_TO_HIGHLIGHT = {
  [TokenType.BOUNDARY]: HighlightSubject.BOUNDARY,
  [TokenType.WORD]: HighlightSubject.WORD,
  [TokenType.BACKTICK_QUOTE]: HighlightSubject.BACKTICK_QUOTE,
  [TokenType.QUOTE]: HighlightSubject.QUOTE,
  [TokenType.NUMBER]: HighlightSubject.NUMBER,
  [TokenType.VARIABLE]: HighlightSubject.VARIABLE,
  [TokenType.LITERAL]: HighlightSubject.LITERAL,
};
