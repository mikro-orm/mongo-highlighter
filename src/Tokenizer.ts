import { TokenType } from './enums';

export interface Token {
  type: number;
  readonly value: string;
}

export class Tokenizer {

  private readonly literal = ['true', 'false', 'null', 'undefined', 'NaN', 'Infinity'];

  /** Punctuation that can be used as a boundary between other tokens */
  private readonly boundaries = [',', ';', ':', ')', '(', '.', '=', '<', '>', '+', '-', '*', '/', '!', '^', '%', '|', '&', '#'];

  // Regular expressions for tokenizing
  private readonly regexBoundaries = '(' + this.quoteRegex(this.boundaries).join('|') + ')';
  private readonly regexLiteral = '(' + this.quoteRegex(this.literal).join('|') + ')';

  /**
   * Takes Mongo query and breaks it into tokens.
   */
  tokenize(string: string): Token[] {
    const tokens: Token[] = [];

    // Used to make sure the string keeps shrinking on each iteration
    let oldStringLen = string.length + 1;
    let token: Token | undefined;
    let currentLength = string.length;

    // Keep processing the string until it is empty
    while (currentLength) {
      // If the string stopped shrinking, there was a problem
      if (oldStringLen <= currentLength) {
        tokens.push({ type: TokenType.ERROR, value: string });
        return tokens;
      }

      oldStringLen = currentLength;

      // Get the next token and the token type
      token = this.createNextToken(string, token);
      const tokenLength = token.value.length;
      tokens.push(token);

      // Advance the string
      string = string.substr(tokenLength);
      currentLength -= tokenLength;
    }

    return tokens;
  }

  /**
   * Return the next token and token type in Mongo query.
   * Quoted strings, comments, reserved words, whitespace, and punctuation are all their own tokens.
   */
  private createNextToken(str: string, previous?: Token): Token {
    let match: RegExpMatchArray | null;

    // Whitespace
    match = str.match(/^\s+/);

    if (match) {
      return { type: TokenType.WHITESPACE, value: match[0] };
    }

    // Quoted String
    if (['"', '\'', '`', '['].includes(str[0])) {
      const type = str[0] === '`' || str[0] === '[' ? TokenType.BACKTICK_QUOTE : TokenType.QUOTE;
      return { type, value: this.getQuotedString(str) };
    }

    // Number (decimal, binary, or hex)
    match = new RegExp(`^([0-9]+(.[0-9]+)?|0x[0-9a-fA-F]+|0b[01]+)($|\\s|"'\`|${this.regexBoundaries})`).exec(str);

    if (match) {
      return { type: TokenType.NUMBER, value: match[1] };
    }

    // Boundary Character (punctuation and symbols)
    match = new RegExp(`^(${this.regexBoundaries})`).exec(str);

    if (match) {
      if (previous && match[1] === ':') {
        previous.type = TokenType.VARIABLE;
      }

      return { type: TokenType.BOUNDARY, value: match[1] };
    }

    const lower = str.toLowerCase();

    // literal
    match = new RegExp(`^(${this.regexLiteral})($|\\s|${this.regexBoundaries})`).exec(lower);

    if (match) {
      return { type: TokenType.LITERAL, value: str.substr(0, match[1].length) };
    }

    // Non reserved word
    match = new RegExp(`^(.*?)($|\\s|["'\`]|${this.regexBoundaries})`).exec(str);

    return { type: TokenType.WORD, value: match![1] };
  }

  /**
   * Helper function for building regular expressions for reserved words and boundary characters
   */
  private quoteRegex(strings: string[]): string[] {
    return strings.map(str => str.replace(new RegExp(`[.\\\\+*?\\[^\\]$(){}=!<>|:\\${'/'}-]`, 'g'), '\\$&'));
  }

  private getQuotedString(string: string): string {
    const re = '^(((`[^`]*($|`))+)|'                    // backtick quoted string using `` to escape
      + '(("[^"\\\\]*(?:\\\\.[^"\\\\]*)*("|$))+)|'      // double quoted string using "" or \" to escape
      + '((\'[^\'\\\\]*(?:\\\\.[^\'\\\\]*)*(\'|$))+))'; // single quoted string using '' or \' to escape
    const m = new RegExp(re, 's').exec(string);

    return m?.[1] ?? '';
  }

}
