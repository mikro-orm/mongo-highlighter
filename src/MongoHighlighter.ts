import c from 'ansi-colors';
import { Token, Tokenizer } from './Tokenizer';
import { TokenType, HighlightSubject, TOKEN_TYPE_TO_HIGHLIGHT } from './enums';

export class MongoHighlighter {

  static readonly DEFAULT_THEME = {
    [HighlightSubject.QUOTE]: c.yellow,
    [HighlightSubject.BACKTICK_QUOTE]: c.yellow,
    [HighlightSubject.BOUNDARY]: c.reset,
    [HighlightSubject.NUMBER]: c.green,
    [HighlightSubject.WORD]: undefined,
    [HighlightSubject.VARIABLE]: c.cyan,
    [HighlightSubject.LITERAL]: c.green,
  };

  private readonly tokenizer = new Tokenizer();

  constructor(private readonly theme: { [K in keyof typeof HighlightSubject]?: string } = {}) {
    this.theme = { ...MongoHighlighter.DEFAULT_THEME, ...this.theme };
  }

  /**
   * Add syntax highlighting to Mongo query
   */
  highlight(query: string): string {
    const tokens = this.tokenizer.tokenize(query);
    let token: Token | undefined;
    let ret = '';
    let position = 0;

    // eslint-disable-next-line no-cond-assign
    while (token = tokens[position++]) {
      ret += this.highlightToken(token.type, token.value);
    }

    return ret;
  }

  private highlightToken(type: TokenType, value: string): string {
    if (type === TokenType.BOUNDARY && ['(', ')'].includes(value)) {
      return value;
    }

    return this.colorize(type, value);
  }

  private colorize(type: TokenType, value: string): string {
    if (!TOKEN_TYPE_TO_HIGHLIGHT[type] || !this.theme[TOKEN_TYPE_TO_HIGHLIGHT[type]]) {
      return value;
    }

    return this.theme[TOKEN_TYPE_TO_HIGHLIGHT[type]](value);
  }

}
