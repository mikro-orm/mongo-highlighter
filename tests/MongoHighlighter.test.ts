import { readFileSync } from 'fs';
import { MongoHighlighter } from '../src';

describe('MongoHighlighter', () => {

  const highlighter = new MongoHighlighter();

  test('highlights query', async () => {
    const queries = readFileSync(__dirname + '/test.txt').toString().split('\n---\n').map(s => s.trim());

    for (const str of queries) {
      const s = highlighter.highlight(str);
      expect(s).toMatchSnapshot(str);
    }
  });

});
