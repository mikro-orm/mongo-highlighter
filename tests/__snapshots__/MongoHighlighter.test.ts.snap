// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoHighlighter highlights query: db.getCollection('author').insertOne({ createdAt: ISODate('2020-01-01T12:00:00Z'), updatedAt: ISODate('2020-01-01T12:00:00Z'), foo: 'bar', name: 'Jon Snow', email: 'snow@wall.st', termsAccepted: false }, { session: '[ClientSession]' }); 1`] = `"db[0m.[0mgetCollection([33m'author'[39m)[0m.[0minsertOne({ [36mcreatedAt[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mupdatedAt[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mfoo[39m[0m:[0m [33m'bar'[39m[0m,[0m [36mname[39m[0m:[0m [33m'Jon Snow'[39m[0m,[0m [36memail[39m[0m:[0m [33m'snow@wall.st'[39m[0m,[0m [36mtermsAccepted[39m[0m:[0m [32mfalse[39m }[0m,[0m { [36msession[39m[0m:[0m [33m'[ClientSession]'[39m })[0m;[0m"`;

exports[`MongoHighlighter highlights query: db.getCollection('book_tag').insert({ createdAt: ISODate('2020-01-01T12:00:00Z'), updatedAt: ISODate('2020-01-01T12:00:00Z'), foo: 'bar', name: 'Jon Snow', email: 'snow@wall.st', termsAccepted: false }, { session: '[ClientSession]' }); 1`] = `"db[0m.[0mgetCollection([33m'book_tag'[39m)[0m.[0minsert({ [36mcreatedAt[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mupdatedAt[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mfoo[39m[0m:[0m [33m'bar'[39m[0m,[0m [36mname[39m[0m:[0m [33m'Jon Snow'[39m[0m,[0m [36memail[39m[0m:[0m [33m'snow@wall.st'[39m[0m,[0m [36mtermsAccepted[39m[0m:[0m [32mfalse[39m }[0m,[0m { [36msession[39m[0m:[0m [33m'[ClientSession]'[39m })[0m;[0m"`;

exports[`MongoHighlighter highlights query: db.getCollection('book_tag').updateMany({ $set: { updatedAt: ISODate('2020-01-01T12:00:00Z'), foo: 123, name: 'Jon Snow', email: 'snow@wall.st', termsAccepted: false } }, { session: '[ClientSession]' }); 1`] = `"db[0m.[0mgetCollection([33m'book_tag'[39m)[0m.[0mupdateMany({ [36m$set[39m[0m:[0m { [36mupdatedAt[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mfoo[39m[0m:[0m [32m123[39m[0m,[0m [36mname[39m[0m:[0m [33m'Jon Snow'[39m[0m,[0m [36memail[39m[0m:[0m [33m'snow@wall.st'[39m[0m,[0m [36mtermsAccepted[39m[0m:[0m [32mfalse[39m } }[0m,[0m { [36msession[39m[0m:[0m [33m'[ClientSession]'[39m })[0m;[0m"`;

exports[`MongoHighlighter highlights query: db.getCollection('book-tag').insertOne({ created_at: ISODate('2020-01-01T12:00:00Z'), meta: { foo: 'bar', baz: 123, lol: true }, foo: 'bar', name: 'Jon Snow', email: 'snow@wall.st', termsAccepted: false }, { session: '[ClientSession]' }); 1`] = `"db[0m.[0mgetCollection([33m'book-tag'[39m)[0m.[0minsertOne({ [36mcreated_at[39m[0m:[0m ISODate([33m'2020-01-01T12:00:00Z'[39m)[0m,[0m [36mmeta[39m[0m:[0m { [36mfoo[39m[0m:[0m [33m'bar'[39m[0m,[0m [36mbaz[39m[0m:[0m [32m123[39m[0m,[0m [36mlol[39m[0m:[0m [32mtrue[39m }[0m,[0m [36mfoo[39m[0m:[0m [33m'bar'[39m[0m,[0m [36mname[39m[0m:[0m [33m'Jon Snow'[39m[0m,[0m [36memail[39m[0m:[0m [33m'snow@wall.st'[39m[0m,[0m [36mtermsAccepted[39m[0m:[0m [32mfalse[39m }[0m,[0m { [36msession[39m[0m:[0m [33m'[ClientSession]'[39m })[0m;[0m"`;
